### DISCUSSION TITLE
Testing on Windows

### DISCUSSION THREAD

**Discussion by mdavis199** [victimsnino/ReactivePlusPlus#194] [General]
Firstly, I"m really excited to see this library and I'm looking forward to trying it out as it develops.  Such great quality and design!

I suppose there is no reason it could not be used on a Windows platform. But, since you are in early development and you seem to be developing on non-windows platforms, I did not want to create a Issue for a problem I'm having while exploring your code.   So, I thought I'd just discuss it here, if you're okay with that.

I've compiled your samples/tests and SGML snake using Microsoft's compiler with no problem.  (Unable to use clang's compiler because LLVM has not yet resolved the MS's ABI incompatibility  with `no_unique_address` attribute.  They should have it resolved in a future version)

I think the Snake code is a great tutorial/overview of its use.  Loved it.  Ran great on windows.

### One Failing Test
All the tests passed, except for `test_group_by.exe`.    It threw an exception for `invalid comparator` at this line of the test;

https://github.com/victimsnino/ReactivePlusPlus/blob/f4ac15e63398fee2f7310984ab89a288c3fc55e1/src/tests/test_group_by.cpp#L252

which ultimately called this line and the `emplace` that crashed:

https://github.com/victimsnino/ReactivePlusPlus/blob/f4ac15e63398fee2f7310984ab89a288c3fc55e1/src/rpp/rpp/operators/group_by.hpp#L111 

The STL source code indicates that this fails because 
```
 // test if _Pred(_Left, _Right) and _Pred is strict weak ordering, when the arguments are the cv-same-type
```

<details> <summary> STL Excerpt</Summary>

```cpp

template <class _Pr, class _Ty1, class _Ty2,
    enable_if_t<is_same_v<_Remove_cvref_t<_Ty1>, _Remove_cvref_t<_Ty2>>, int> = 0>
constexpr bool _Debug_lt_pred(_Pr&& _Pred, _Ty1&& _Left, _Ty2&& _Right) noexcept(
    noexcept(_Pred(_Left, _Right)) && noexcept(_Pred(_Right, _Left))) {
    // test if _Pred(_Left, _Right) and _Pred is strict weak ordering, when the arguments are the cv-same-type
    const auto _Result = static_cast<bool>(_Pred(_Left, _Right));
    if (_Result) {
        _STL_VERIFY(!_Pred(_Right, _Left), "invalid comparator");
    }

    return _Result;
}

```
</details>

That's as far as I've gone on drilling down into the problem.  I just wanted to give you a heads up.

**Comment by AlexInLog** [comment]
Hi @mdavis199, thanks for your post! And you are welcome =)

> you seem to be developing on non-windows platforms

No, my primary development environment is windows =) 


Thanks for your point and feedback! =)


About failing test: sounds interesting because I can't see such a problem locally. Also CI on Windows also works [fine](https://github.com/victimsnino/ReactivePlusPlus/runs/7560777073?check_suite_focus=true).

Looks like you've been run tests in Debug build (as a result, all `_STL_VERIFY` are enabled)



Can you check if Release build works as expected for you? 

Anyway, this error makes sense, I will try to fix it


UPD: Also feel free to submit issues =)

**Comment by AlexInLog** [comment]
UPD: reproduced and found issue related to this one. Thanks for it! Can you create issue related to this one?

**Comment by mdavis199** [comment]
Yes, you're right, of course.   All tests pass in release mode.    Sure, I'll create the issue.  Thanks!
